### PR TITLE
fallback to /dev/mapper device if metadata device is not set in docker info

### LINKED ISSUE #1338: cadvisor does not return any container stats when using devicemapper with no metadata file
docker-1.10 is using thin pool that is created by docker-storage-setup on rhel7, and in this case, metadata file is not populated:

```
#docker info
Containers: 6
 Running: 2
 Paused: 0
 Stopped: 4
Images: 39
Server Version: 1.10.3
Storage Driver: devicemapper
 Pool Name: rhel-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 Data Space Used: 1.898 GB
 Data Space Total: 8.565 GB
 Data Space Available: 6.667 GB
 Metadata Space Used: 524.3 kB
 Metadata Space Total: 54.53 MB
 Metadata Space Available: 54 MB
 Udev Sync Supported: true
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 4
 Library Version: 1.02.124-RHEL7 (2016-05-04)
Execution Driver: native-0.2
Logging Driver: journald
Plugins: 
 Volume: local
 Network: bridge null host
 Authorization: rhel-push-plugin
Kernel Version: 3.10.0-426.el7.x86_64
Operating System: Red Hat Enterprise Linux
OSType: linux
Architecture: x86_64
Number of Docker Hooks: 2
CPUs: 4
Total Memory: 7.64 GiB
Name: localhost.localdomain
ID: VFQS:UYX7:SQLV:5QOP:M3XL:7TA3:GG7I:PUDX:L2E5:A6YO:MLCY:GLN4
WARNING: bridge-nf-call-ip6tables is disabled
Registries: registry.access.redhat.com (secure), docker.io (secure)
```

Now when running a kube cluster, one could see the following error:

```
Docker container factory registration failed: couldn't determine devicemapper metadata device: Could not get the devicemapper metadata device.
```

The result is that getting stats on a local cluster fails for pods:

```
curl http://127.0.0.1:10255/stats/summary 
 ```
It only shows empty `"pods": []`, even though there is a pod and its containers.

And also getting stats for a specific container also fails:

````
curl http://127.0.0.1:10255/stats/<pod-name>/<container-name>

Internal Error: unable to find Docker container "021a9d9bc5c44d5c052b78f13079e1259c4fc76e633657386a745b9edfc5bc47"
```

### PR DESCRIPTION
**Pull request by sjenning**
@derekwaynecarr @aveshagarwal @pmorie @ncdc 

In systems that have an LVM managed docker thin pool, docker info does not contain information for the data or metadata devices:

```
$ docker info
..
Storage Driver: devicemapper
 Pool Name: fedora-docker--pool
 Pool Blocksize: 524.3 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: 
 Metadata file: 
 ...
```

LVM puts the metadata device at /dev/mapper/<poolname>_tmeta.  If we fail to get the metadata device from the docker info, we should fall back to the LVM path.

Fixes #1338

### CODE REVIEW

**Comment by ncdc** [comment]
FYI @vishh @dchen1107 @timstclair the initial "thin_ls" PR that we added only works for devicemapper-managed thin pools. This PR should allow support for LVM-managed pools as well. AFAIK this PR helps us pass some node E2E tests that are currently failing on the Red Hat stack.

cc @rhvgoyal @ingvagabund

**Comment by ncdc** [comment]
If this isn't approved for Kube 1.3, we'd love to see it ASAP after.

**Comment by sjenning** [comment]
Just added some new changes
- os.Stat to check to metadata file existence when using assumed LVM path
- only fail to gather filesystem stats if metadata device can't be found (other stats still work)

File: utils/docker/docker.go [hunk 1]
```diff
@@ -50,8 +51,20 @@ func DockerThinPoolName(info dockertypes.Info) (string, error) {
...
+	}
+
+	poolName, err := DockerThinPoolName(info)
+	if err != nil {
+		return "", err
+	}
+
+	metadataDevice = fmt.Sprintf("/dev/mapper/%s_tmeta", poolName)
+
+	_, err = os.Stat(metadataDevice)
```
Comment (ncdc) [inline utils/docker/docker.go]: if _, err := os.Stat(metadataDevice); err != nil {

File: container/docker/factory.go [hunk 2]
```diff
@@ -209,15 +209,16 @@ func Register(factory info.MachineInfoFactory, fsInfo fs.FsInfo, ignoreMetrics c
 
 			dockerMetadataDevice, err := dockerutil.DockerMetadataDevice(*dockerInfo)
 			if err != nil {
-				return fmt.Errorf("couldn't determine devicemapper metadata device: %v", err)
+				glog.Errorf("couldn't determine devicemapper metadata device: %v", err)
+				glog.Errorf("devicemapper filesystem stats will not be reported")
```
Comment (ncdc) [inline container/docker/factory.go]: I would only do one Errorf in this case - can you combine this with the previous one?

**Comment by vishh** [comment]
LGTM. Once the existing comments are addressed, this is good to go. @ncdc is this supposed to be integrated into kube v1.3?

**Comment by ncdc** [comment]
@vishh if this can make 1.3, that would be great

**Comment by vishh** [comment]
That would require a cherrypick against the `release-0.23` branch. A new
tag needs to be cut against v0.23 and then vendor that tag into kube.

**Comment by ncdc** [comment]
@vishh assuming this gets merged, we can do the PR against release-0.23 and then update the kube godep once you guys cut a new tag

**Comment by derekwaynecarr** [comment]
I think this needs to make 1.3 as without it pod stats are not reported at all.

**Comment by ncdc** [comment]
I think this looks good. @vishh can you kick the e2e?

**Comment by ncdc** [comment]
Not sure if I have admin, but let's see:

@k8s-bot ok to test

**Comment by timstclair** [comment]
ok to test

**Comment by vishh** [comment]
@k8s-bot test this

**Comment by timstclair** [comment]
CoreOS issue: https://github.com/google/cadvisor/issues/1344 - looks like this has started affecting the build job as well.

**Comment by timstclair** [comment]
Disabled coreos-beta.

@k8s-bot test this

**Comment by ncdc** [comment]
@vishh @derekwaynecarr any other comments?

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by ncdc** [comment]
I'd like to get this merged ASAP once it passes reviews so we can prep the follow-on tasks (release-0.23 cherry-pick, kube godep bump)... any other comments?

**Comment by vishh** [comment]
LGTM

**Pull request merged**